## 自然言語処理 期末試験（実習分）

問1〜5のソースコードと実行結果を示し、また、問に記述解答する箇所は入力し完成させてください。**必要なセルは適宜挿入しながら進めてください。コード以外の記述解答はmarkdownで記入し実行しましょう。ファイル名の00とnameは自分の出席番号と名前（アルファベット:nobi_nobita)に変更してください．**<br>
以下の番号名前を記入してから始めてください。  

<font color="MediumVioletRed">\[必須\] </font>番号：08　　　　　氏名：イマムカイリルビス　　　　　　　　　　　　　　　　　　　　　　　　　　　点数：

---

### 問1

\[形態素解析\]<br>

1日本語を入力するとウチュウ（宇宙）語に翻訳するコードを作成する問題です．  
　ウチュウ語に翻訳する規則を次のように設定します．   

**規則：** すべてをカタカナに変換し，「助詞」を取り除く

 問題文中の"l"は行番号を表す記号を表します．

l14:「ヨミガナを返す関数にtokenを渡しヨミガナを取得」, l15:「ウチュウ語の結果を返すように書く」，l19:「適切な品詞を書く」を適切なコードに書き換えて日本語（一文）をウチュウ語（一文）に変換する関数nihongo2ucyugo()と単語のヨミガナを返す関数token2ucyugo()を定義し，以下の文をnihongo2ucyugo()を用い連続技で翻訳せよ．  

1. われわれは地球を救うためにエム78星雲からやってきた。
2. 警察は自転車で逃げる泥棒を追いかけた。
3. 私をベクトルにできる者はおるか。

以下のコードを修正し，解答してください．

In [ ]:
!pip install janome

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.7/19.7 MB 67.0 MB/s eta 0:00:00


In [ ]:
#形態素解析にはjanomeを用います
from janome.tokenizer import Tokenizer

# Tokenizerインスタンスの生成 
tokenizer = Tokenizer()

# 日本語をウチュウにする関数
def nihongo2ucyugo(input_str):
  # 形態素解析の実施
  tokens = tokenizer.tokenize(input_str)
  # 各token の変換結果を" "（半角スペース) でつなげる
  result_str = ""
  for token in tokens:
    result_str += token2ucyugo(token) + " "
  return result_str

# tokenが助詞の場合は空文字列、それ以外はヨミガナを返す関数
def token2ucyugo(input_token):
  if input_token.part_of_speech.split(',')[0] == "助詞":
    return ""
  else:
    return input_token.reading

text = ["われわれは地球を救うためにエム78星雲からやってきた。",
        "警察は自転車で逃げる泥棒を追いかけた。",
        "私をベクトルにできる者はおるか。"]

#動作確認のプリント文挿入
print(nihongo2ucyugo(text[0]))

ワレワレ  チキュウ  スクウ タメ  * * セイウン  ヤッテキ タ 。 


In [ ]:
#連続技で実行
for sentence in text:
  print(nihongo2ucyugo(sentence))

ワレワレ  チキュウ  スクウ タメ  * * セイウン  ヤッテキ タ 。 
ケイサツ  ジテンシャ  ニゲル ドロボウ  オイカケ タ 。 
ワタシ  ベクトル  デキル モノ  オル  。 


#### ウチュウ語翻訳結果を以下に記述してください．
1. ワレワレ  チキュウ  スクウ タメ  * * セイウン  ヤッテキ タ 。 
2. ケイサツ  ジテンシャ  ニゲル ドロボウ  オイカケ タ 。 
3. ワタシ  ベクトル  デキル モノ  オル  。 

<div style="text-align: right;">
[]点
</div>

---

## 問2

\[正規表現， 単語カウント\]<br>
小説のテキストファイルに対しクリーニング処理を行い，単語出現頻度を求める問題です．以下の1~3の設問に沿って解答してください． また，4,5の問に解答してください．

1. 青空文庫「怪人二十面相(かいじんにじゅうめんそう」: kaijin_nijumenso.txt に対して、ルビ、読みの注意、"|" と全角半角スペース、"「」"と改行を正規表現を用いて削除し、テキストの前処理を行い、['文1', '文2', '文3', ... , '文 N'] のリストをpickle形式で保存せよ。<br>
読み込んだ kaijin_nijumenso.txtのテキストデータと前処理後のリスト形式のデータを表示し，前処理がされているか確認すること。
printで不具合がある場合は，import pprintを追加し， printをpprintに変更するなど対応してください。<br>

In [ ]:
with open("kaijin_nijumenso.txt", mode = "r", encoding = "utf-8") as f:
  kaijin_original = f.read()

# print(kaijin_original)

In [ ]:
import re

kaijin = re.sub("《[^》]+》", "", kaijin_original) 
kaijin = re.sub("［[^］]+］", "", kaijin) 
kaijin = re.sub("[｜ 　「」\n]", "", kaijin)

# print(kaijin)

In [ ]:
separator = "。"

kaijin_list = kaijin.split(separator)
kaijin_list.pop()
kaijin_list = [x + separator for x in kaijin_list]

print(kaijin_list)

['そのころ、東京中の町という町、家という家では、ふたり以上の人が顔をあわせさえすれば、まるでお天気のあいさつでもするように、怪人二十面相のうわさをしていました。', '二十面相というのは、毎日毎日、新聞記事をにぎわしている、ふしぎな盗賊のあだ名です。', 'その賊は二十のまったくちがった顔を持っているといわれていました。', 'つまり、変装がとびきりじょうずなのです。', 'どんなに明るい場所で、どんなに近よってながめても、少しも変装とはわからない、まるでちがった人に見えるのだそうです。', '老人にも若者にも、富豪にも乞食にも、学者にも無頼漢にも、いや、女にさえも、まったくその人になりきってしまうことができるといいます。', 'では、その賊のほんとうの年はいくつで、どんな顔をしているのかというと、それは、だれひとり見たことがありません。', '二十種もの顔を持っているけれど、そのうちの、どれがほんとうの顔なのだか、だれも知らない。', 'いや、賊自身でも、ほんとうの顔をわすれてしまっているのかもしれません。', 'それほど、たえずちがった顔、ちがった姿で、人の前にあらわれるのです。', 'そういう変装の天才みたいな賊だものですから、警察でもこまってしまいました。', 'いったい、どの顔を目あてに捜索したらいいのか、まるで見当がつかないからです。', 'ただ、せめてものしあわせは、この盗賊は、宝石だとか、美術品だとか、美しくてめずらしくて、ひじょうに高価な品物をぬすむばかりで、現金にはあまり興味を持たないようですし、それに、人を傷つけたり殺したりする、ざんこくなふるまいは、一度もしたことがありません。', '血がきらいなのです。', 'しかし、いくら血がきらいだからといって、悪いことをするやつのことですから、自分の身があぶないとなれば、それをのがれるためには、何をするかわかったものではありません。', '東京中の人が二十面相のうわさばかりしているというのも、じつは、こわくてしかたがないからです。', 'ことに、日本にいくつという貴重な品物を持っている富豪などは、ふるえあがってこわがっていました。', '今までのようすで見ますと、いくら警察へたのんでも、ふせぎようのない、おそろしい賊なのですから。', 'この二十面相には、一つのみょうなくせがありました。', '何か

In [ ]:
import pickle

with open("kaijin_list.pickle", mode ="wb") as f:
  pickle.dump(kaijin_list, f)

  2. 1で生成したpickleファイルを読み込み、内容を表示する。

In [ ]:
with open("kaijin_list.pickle", mode = "rb") as f:
  kaijin_list = pickle.load(f)

print(kaijin_list)

['そのころ、東京中の町という町、家という家では、ふたり以上の人が顔をあわせさえすれば、まるでお天気のあいさつでもするように、怪人二十面相のうわさをしていました。', '二十面相というのは、毎日毎日、新聞記事をにぎわしている、ふしぎな盗賊のあだ名です。', 'その賊は二十のまったくちがった顔を持っているといわれていました。', 'つまり、変装がとびきりじょうずなのです。', 'どんなに明るい場所で、どんなに近よってながめても、少しも変装とはわからない、まるでちがった人に見えるのだそうです。', '老人にも若者にも、富豪にも乞食にも、学者にも無頼漢にも、いや、女にさえも、まったくその人になりきってしまうことができるといいます。', 'では、その賊のほんとうの年はいくつで、どんな顔をしているのかというと、それは、だれひとり見たことがありません。', '二十種もの顔を持っているけれど、そのうちの、どれがほんとうの顔なのだか、だれも知らない。', 'いや、賊自身でも、ほんとうの顔をわすれてしまっているのかもしれません。', 'それほど、たえずちがった顔、ちがった姿で、人の前にあらわれるのです。', 'そういう変装の天才みたいな賊だものですから、警察でもこまってしまいました。', 'いったい、どの顔を目あてに捜索したらいいのか、まるで見当がつかないからです。', 'ただ、せめてものしあわせは、この盗賊は、宝石だとか、美術品だとか、美しくてめずらしくて、ひじょうに高価な品物をぬすむばかりで、現金にはあまり興味を持たないようですし、それに、人を傷つけたり殺したりする、ざんこくなふるまいは、一度もしたことがありません。', '血がきらいなのです。', 'しかし、いくら血がきらいだからといって、悪いことをするやつのことですから、自分の身があぶないとなれば、それをのがれるためには、何をするかわかったものではありません。', '東京中の人が二十面相のうわさばかりしているというのも、じつは、こわくてしかたがないからです。', 'ことに、日本にいくつという貴重な品物を持っている富豪などは、ふるえあがってこわがっていました。', '今までのようすで見ますと、いくら警察へたのんでも、ふせぎようのない、おそろしい賊なのですから。', 'この二十面相には、一つのみょうなくせがありました。', '何か

3. 1で生成したpikleファイルを読み込み、小説「怪人二十面相」に出現する単語をカウントし表示する。

In [ ]:
t = Tokenizer()

word = []

for sentence in kaijin_list:
  word += t.tokenize(sentence, wakati=True)

print(word)

['その', 'ころ', '、', '東京', '中', 'の', '町', 'という', '町', '、', '家', 'という', '家', 'で', 'は', '、', 'ふたり', '以上', 'の', '人', 'が', '顔', 'を', 'あわせ', 'さえ', 'すれ', 'ば', '、', 'まるで', 'お', '天気', 'の', 'あいさつ', 'でも', 'する', 'よう', 'に', '、', '怪人', '二', '十', '面相', 'の', 'うわさ', 'を', 'し', 'て', 'い', 'まし', 'た', '。', '二', '十', '面相', 'という', 'の', 'は', '、', '毎日', '毎日', '、', '新聞', '記事', 'を', 'にぎわし', 'て', 'いる', '、', 'ふしぎ', 'な', '盗賊', 'の', 'あだ名', 'です', '。', 'その', '賊', 'は', '二', '十', 'の', 'まったく', 'ちがっ', 'た', '顔', 'を', '持っ', 'て', 'いる', 'と', 'いわ', 'れ', 'て', 'い', 'まし', 'た', '。', 'つまり', '、', '変装', 'が', 'とびきり', 'じょうず', 'な', 'の', 'です', '。', 'どんなに', '明るい', '場所', 'で', '、', 'どんなに', '近', 'よって', 'ながめ', 'て', 'も', '、', '少し', 'も', '変装', 'と', 'は', 'わから', 'ない', '、', 'まるで', 'ちがっ', 'た', '人', 'に', '見える', 'の', 'だ', 'そう', 'です', '。', '老人', 'に', 'も', '若者', 'に', 'も', '、', '富豪', 'に', 'も', '乞食', 'に', 'も', '、', '学者', 'に', 'も', '無頼漢', 'に', 'も', '、', 'いや', '、', '女', 'に', 'さえ', 'も', '、', 'まったく', 'その', '人', 'に', 'なり', 'きっ', 'て', 'しまう', 'こと', 'が', 'できる', 'と'

In [26]:
import collections

c = collections.Counter(word)

print(c.most_common())

[('、', 5620), ('の', 3463), ('。', 3164), ('て', 2203), ('た', 2139), ('に', 1948), ('は', 1890), ('を', 1734), ('が', 1507), ('と', 1052), ('です', 904), ('で', 698), ('まし', 693), ('し', 666), ('も', 653), ('ん', 583), ('か', 489), ('ない', 457), ('な', 446), ('だ', 417), ('い', 406), ('から', 390), ('二', 379), ('こと', 356), ('よう', 346), ('ませ', 341), ('十', 335), ('いる', 334), ('ます', 302), ('その', 300), ('明智', 273), ('う', 270), ('へ', 266), ('あり', 252), ('面相', 246), ('でし', 242), ('ぼく', 213), ('賊', 210), ('…', 202), ('よ', 198), ('ね', 194), ('この', 179), ('れ', 178), ('君', 169), ('という', 168), ('もの', 168), ('探偵', 168), ('それ', 165), ('人', 159), ('ある', 156), ('きみ', 152), ('中', 144), ('そう', 143), ('ば', 142), ('さ', 141), ('き', 128), ('お', 127), ('する', 127), ('じゃ', 124), ('見', 123), ('壮', 119), ('小林', 117), ('手', 117), ('少年', 116), ('顔', 113), ('でしょ', 113), ('老人', 110), ('なっ', 110), ('まで', 100), ('一', 94), ('ながら', 93), ('いっ', 92), ('何', 89), ('もう', 88), ('これ', 87), ('名', 87), ('られ', 85), ('氏', 84), ('三', 84), ('ほど', 84), 

4. 単語  "の"、"が"、"明智"、および、"探偵"のそれぞれの出現頻度を答えよ。
#### 単語出現頻度
の：  3463<br>
が：  1507<br>
明智：  273<br>
探偵：  168

In [27]:
print(c.get("の"))
print(c.get("が"))
print(c.get("明智"))
print(c.get("探偵"))

3463
1507
273
168


5. 小説の特徴を分析する際に，4で解答した"の"と "が" は"明智"や"探偵"より頻出していますがこれらは特徴を表しているといえるでしょうか？ 適切な処理と思う選択肢の記号を選んでください．（複数回答可）　　

A. あらかじめ助詞を取り除く前処理を行う  
B. 不要な語はストップワードリストを作成しそのリストにある語を取り除いてから出現頻度を調べる  
C. 元のテキストデータから，手作業で不要な語を一個一個削除する  
D. そのまま残す

**解答：**A, B


<div style="text-align: right;">
[]点
</div>

---

## 問3
\[文章ベクトル\]<br>
小説「怪人二十面相」のテキストをルビや記号を削除する前処理し(week13）pickleファイルにした後、Doc2vecを用いた文章の分散表現（学習済みモデル）をDBOWで生成し、類似文章を求めよ。以下の1~6の手順に従って解答してください。また7,8の問に解答してください。(Word2vecではないことに注意）<br>
文章ベクトル（分散表現）生成のパラメータは実習のときと同じ設定とします。

1. テキストの前処理をし、[[1文の形態素リスト], [1文の形態素リスト], [1文の形態素リスト], ...] の形式でpickle形式に保存する。

In [29]:
sentence_list = [[sentence] for sentence in kaijin_list]

print(sentence_list)

[['そのころ、東京中の町という町、家という家では、ふたり以上の人が顔をあわせさえすれば、まるでお天気のあいさつでもするように、怪人二十面相のうわさをしていました。'], ['二十面相というのは、毎日毎日、新聞記事をにぎわしている、ふしぎな盗賊のあだ名です。'], ['その賊は二十のまったくちがった顔を持っているといわれていました。'], ['つまり、変装がとびきりじょうずなのです。'], ['どんなに明るい場所で、どんなに近よってながめても、少しも変装とはわからない、まるでちがった人に見えるのだそうです。'], ['老人にも若者にも、富豪にも乞食にも、学者にも無頼漢にも、いや、女にさえも、まったくその人になりきってしまうことができるといいます。'], ['では、その賊のほんとうの年はいくつで、どんな顔をしているのかというと、それは、だれひとり見たことがありません。'], ['二十種もの顔を持っているけれど、そのうちの、どれがほんとうの顔なのだか、だれも知らない。'], ['いや、賊自身でも、ほんとうの顔をわすれてしまっているのかもしれません。'], ['それほど、たえずちがった顔、ちがった姿で、人の前にあらわれるのです。'], ['そういう変装の天才みたいな賊だものですから、警察でもこまってしまいました。'], ['いったい、どの顔を目あてに捜索したらいいのか、まるで見当がつかないからです。'], ['ただ、せめてものしあわせは、この盗賊は、宝石だとか、美術品だとか、美しくてめずらしくて、ひじょうに高価な品物をぬすむばかりで、現金にはあまり興味を持たないようですし、それに、人を傷つけたり殺したりする、ざんこくなふるまいは、一度もしたことがありません。'], ['血がきらいなのです。'], ['しかし、いくら血がきらいだからといって、悪いことをするやつのことですから、自分の身があぶないとなれば、それをのがれるためには、何をするかわかったものではありません。'], ['東京中の人が二十面相のうわさばかりしているというのも、じつは、こわくてしかたがないからです。'], ['ことに、日本にいくつという貴重な品物を持っている富豪などは、ふるえあがってこわがっていました。'], ['今までのようすで見ますと、いくら警察へたのんでも、ふせぎようのない、おそろしい賊なのですから。

In [32]:
with open("sentence_list.pickle", mode = "wb") as f:
  pickle.dump(sentence_list, f)

In [46]:
with open("sentence_list.pickle", mode = "rb") as f:
  sentence = pickle.load(f)

print(sentence)

[['そのころ、東京中の町という町、家という家では、ふたり以上の人が顔をあわせさえすれば、まるでお天気のあいさつでもするように、怪人二十面相のうわさをしていました。'], ['二十面相というのは、毎日毎日、新聞記事をにぎわしている、ふしぎな盗賊のあだ名です。'], ['その賊は二十のまったくちがった顔を持っているといわれていました。'], ['つまり、変装がとびきりじょうずなのです。'], ['どんなに明るい場所で、どんなに近よってながめても、少しも変装とはわからない、まるでちがった人に見えるのだそうです。'], ['老人にも若者にも、富豪にも乞食にも、学者にも無頼漢にも、いや、女にさえも、まったくその人になりきってしまうことができるといいます。'], ['では、その賊のほんとうの年はいくつで、どんな顔をしているのかというと、それは、だれひとり見たことがありません。'], ['二十種もの顔を持っているけれど、そのうちの、どれがほんとうの顔なのだか、だれも知らない。'], ['いや、賊自身でも、ほんとうの顔をわすれてしまっているのかもしれません。'], ['それほど、たえずちがった顔、ちがった姿で、人の前にあらわれるのです。'], ['そういう変装の天才みたいな賊だものですから、警察でもこまってしまいました。'], ['いったい、どの顔を目あてに捜索したらいいのか、まるで見当がつかないからです。'], ['ただ、せめてものしあわせは、この盗賊は、宝石だとか、美術品だとか、美しくてめずらしくて、ひじょうに高価な品物をぬすむばかりで、現金にはあまり興味を持たないようですし、それに、人を傷つけたり殺したりする、ざんこくなふるまいは、一度もしたことがありません。'], ['血がきらいなのです。'], ['しかし、いくら血がきらいだからといって、悪いことをするやつのことですから、自分の身があぶないとなれば、それをのがれるためには、何をするかわかったものではありません。'], ['東京中の人が二十面相のうわさばかりしているというのも、じつは、こわくてしかたがないからです。'], ['ことに、日本にいくつという貴重な品物を持っている富豪などは、ふるえあがってこわがっていました。'], ['今までのようすで見ますと、いくら警察へたのんでも、ふせぎようのない、おそろしい賊なのですから。

2. Doc2vecを用い文章の分散表現を生成する。（手法はDBOW→ 0）

In [53]:
lenght = len(sentence)

sentence_new = []

for i in range(lenght):
    words = []
    for token in t.tokenize(sentence[i][0], wakati=True):
        words.append(token)
    sentence_new.append(words)

In [54]:
print(sentence_new)

[['その', 'ころ', '、', '東京', '中', 'の', '町', 'という', '町', '、', '家', 'という', '家', 'で', 'は', '、', 'ふたり', '以上', 'の', '人', 'が', '顔', 'を', 'あわせ', 'さえ', 'すれ', 'ば', '、', 'まるで', 'お', '天気', 'の', 'あいさつ', 'でも', 'する', 'よう', 'に', '、', '怪人', '二', '十', '面相', 'の', 'うわさ', 'を', 'し', 'て', 'い', 'まし', 'た', '。'], ['二', '十', '面相', 'という', 'の', 'は', '、', '毎日', '毎日', '、', '新聞', '記事', 'を', 'にぎわし', 'て', 'いる', '、', 'ふしぎ', 'な', '盗賊', 'の', 'あだ名', 'です', '。'], ['その', '賊', 'は', '二', '十', 'の', 'まったく', 'ちがっ', 'た', '顔', 'を', '持っ', 'て', 'いる', 'と', 'いわ', 'れ', 'て', 'い', 'まし', 'た', '。'], ['つまり', '、', '変装', 'が', 'とびきり', 'じょうず', 'な', 'の', 'です', '。'], ['どんなに', '明るい', '場所', 'で', '、', 'どんなに', '近', 'よって', 'ながめ', 'て', 'も', '、', '少し', 'も', '変装', 'と', 'は', 'わから', 'ない', '、', 'まるで', 'ちがっ', 'た', '人', 'に', '見える', 'の', 'だ', 'そう', 'です', '。'], ['老人', 'に', 'も', '若者', 'に', 'も', '、', '富豪', 'に', 'も', '乞食', 'に', 'も', '、', '学者', 'に', 'も', '無頼漢', 'に', 'も', '、', 'いや', '、', '女', 'に', 'さえ', 'も', '、', 'まったく', 'その', '人', 'に', 'なり', 'きっ', 'て', 'しまう', 'こと', 'が',

In [58]:
from gensim.models.doc2vec import Doc2Vec
from gensim.models.doc2vec import TaggedDocument

tagged_documents = []
for i, s in enumerate(sentence_new):
    tagged_documents.append(TaggedDocument(s, [i]))

model = Doc2Vec(documents=tagged_documents,
                vector_size=100,
                min_count=5,
                window=5,
                epochs=20,
                dm=0)

3. リストの2946番目の文章とベクトルを表示

In [60]:
print(model.docvecs[2946])

[ 0.01979873  0.05001161  0.04243043 -0.01182222  0.02557759 -0.02614273
 -0.06716198  0.03752917 -0.11705827  0.2164293  -0.10929488 -0.00240958
 -0.00084052  0.04661932 -0.06369887  0.12192888 -0.10022452  0.14688234
 -0.05338409  0.02804417 -0.03436763 -0.13933127 -0.02271739  0.20644808
 -0.05318919 -0.0588726   0.08907373 -0.05272    -0.04430884  0.11516188
 -0.07359263 -0.11018346  0.11989606 -0.17285703 -0.0151512  -0.04316171
  0.03380177 -0.04567462 -0.00265879 -0.06874464  0.20062883  0.02555409
 -0.05161356 -0.15974316  0.0937585   0.02072214 -0.00657055 -0.06192387
 -0.12650536 -0.04284955 -0.02253103 -0.12455214  0.04283927  0.03473016
 -0.1305144   0.06468908 -0.01780137  0.0501552   0.17527378 -0.00642815
 -0.14308083  0.01137661 -0.04543716  0.22461705  0.02793904 -0.03720054
  0.15796301 -0.07881432 -0.00804596 -0.09827783  0.007846   -0.02217899
  0.10565638 -0.10963199  0.05131365 -0.16604249  0.09121031  0.16126433
 -0.05987073  0.00861832  0.08813962  0.02039486  0

4. 2946番目の文章と類似度の高い文章のIDと類似度を表示（デフォルトの降順10個）

In [64]:
for p in model.docvecs.most_similar(2946):
  print(p)

(1867, 0.9991791248321533)
(2487, 0.9990788102149963)
(607, 0.999069333076477)
(1636, 0.9990618228912354)
(3161, 0.9990249276161194)
(1697, 0.9990198016166687)
(1613, 0.9989981651306152)
(604, 0.9989923238754272)
(323, 0.9989913702011108)
(1432, 0.9989904761314392)


5. 2946番目の文章と類似度の高い文章を10文表示（デフォルトの降順10個）

In [68]:
for p in model.docvecs.most_similar(2946):
  print(sentence[p[0]])

['明智は平然として、このおどろくべきことばを語りました。']
['そのねらわれた人物というのは、ああ、やっぱり明智探偵でした。']
['おとうさま、それは明智小五郎探偵よ。']
['明智さん、これはなんということです。']
['そして一同、かわいらしい声をそろえて、くりかえしくりかえしさけぶのでした。']
['明智はすまして、いよいよへんなことをいうのです。']
['明智は刑事の肩をつかんで、らんぼうにゆさぶるのでした。']
['しかし、個人の探偵などにこの大事件がこなせるかしらん。']
['同君の家出以前の写真も手に入れました。']
['あの方が、明智先生でございます。']


6. 4の結果から，最も類似度が高い文章IDを入力し，その文章を表示（1文）

In [73]:
print(sentence[1867])

['明智は平然として、このおどろくべきことばを語りました。']


7. このDoc2Vecモデルの文書ベクトルの次元は何次元で生成しましたか？
**解答：**  [  100   ] 次元

8. ID2946と最も類似度が高い文書IDと文を答えてください．
**解答：**  
文書ID:  1867<br>
文：明智は平然として、このおどろくべきことばを語りました。

<div style="text-align: right;">
[]点
</div>

---

## 問4
\[評価値\]<br>
映画評価コメントをポジティブな意見かネガティブな意見をまとめたデータに対してロジスティック回帰モデルを適用し，正解のラベル，予測されたラベル，予測確率をタブ区切り形式で出力したファイル result.txt を利用し，その予測結果を次の評価値で評価せよ。<br>
1.正解率，2.適合率，3.再現率および 4.F値

In [75]:
fname_result = 'result.txt' 

def score(fname):
    '''結果ファイルからスコア算出
    結果ファイルを読み込んで、正解率、適合率、再現率、F1スコアを返す
    戻り値：
    正解率,適合率,再現率,F1スコア
    '''
    TP = 0      # True-Positive     予想が+1、正解も+1
    FP = 0      # False-Positive    予想が+1、正解は-1
    FN = 0      # False-Negative    予想が-1、正解は+1
    TN = 0      # True-Negative     予想が-1、正解も-1

    with open(fname) as data_file: #ファイルをopenし，data_fileとする
        for line in data_file:
            cols = line.split('\t')    # "-1  -1  0.7992886809287588"の並びのデータをtabで分割しリスト化
#            print(cols)   #colsの内容を表示

            if len(cols) < 3:
                continue

            if cols[0] == '+1':         # 0番目：正解極性
                if cols[1] == '+1':     # 1番目：予想極性
                    TP += 1
                else:
                    FN += 1
            else:
                if cols[1] == '+1':
                    FP += 1
                else:
                    TN += 1

    accuracy = (TP + TN) / (TP + FN + FP + TN)    # 正解率
    precision = TP / (TP+FP)     # 適合率
    recall = TP/(TP+FN)    # 再現率
    f1 = (2*precision*recall) / (recall + precision)   # F1スコア

    return accuracy, precision, recall, f1

accuracy, precision, recall, f1 = score(fname_result)
print('正解率　\t{}\n適合率　\t{}\n再現率　\t{}\nF1スコア　\t{}'.format( accuracy, precision, recall, f1))

正解率　	0.7477485928705441
適合率　	0.7476092255765986
再現率　	0.748030018761726
F1スコア　	0.7478195629747725


**問：**　この結果について，考察してください． 「〇〇のことから，△△といえる」のような根拠から導出される結論の文で答えること。この文型に合わせなくても根拠からの結論が記述されていれば構わない。



**解答：**
正確率は 70％ 超えていることから、このデータは結構正確であることといえる。90％ぐらいの正確率よりあまり正確とも言えない。

<div style="text-align: right;">
[]点
</div>

---

## 問5

\[単語の重要度\]<br>
以下の文書0〜3の4文を用いて，各素性（単語）のTF-IDFを求めよ。TF-IDFをpandasライブラリを利用してデータフレームを用い表示し<span style="color: skyblue; ">（最終的に，そのほかの配列や値は表示する必要はない．）</span>，文書0~3の特徴のある単語( 𝑡𝑓⋅𝑖𝑑𝑓>0.45 とした場合)を答えてください。複数ある場合は，カンマで区切ること。<br>
手法はTF-IDFの式を直接書く方法でもTfidfVectorizerを用いる方法でもどちらを選択しても構わない．

- 0: 'This is the first document.'
- 1: 'This document is the second document.'
- 2: 'And this is the third one.'
- 3: 'Is this the first document?'

In [85]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import normalize #データの正規化などが可能な
import numpy as np #数値計算用ライブラリpnumpyをnpと名前をつけて以下で利用
import pandas as pd #データ解析用ライブラリpandasをpdと名前をつけて以下で利用

corpus = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?'
]

smooth_idf = True
norm_idf = True

wc = CountVectorizer()
x = wc.fit_transform(corpus)
wcX = np.array(x.toarray())

print('4文コーパス: \n', corpus, '\n') 
print('頻度ベクトル: \n', wcX,'\n')  

4文コーパス: 
 ['This is the first document.', 'This document is the second document.', 'And this is the third one.', 'Is this the first document?'] 

頻度ベクトル: 
 [[0 1 1 1 0 0 1 0 1]
 [0 2 0 1 0 1 1 0 1]
 [1 0 0 1 1 0 1 1 1]
 [0 1 1 1 0 0 1 0 1]] 



In [84]:
N = wcX.shape[0]

tf = np.array([wcX[i, :] / np.sum(wcX, axis=1)[i] for i in range(N)])

print('4文コーパス: \n', corpus, '\n') 
print('全文書数： ', N, '\n') 


print('wcX[i, :] \n', [wcX[i, :]for i in range(N)], '\n')
print('words_freq \n', [np.sum(wcX, axis=1)[i]for i in range(N)], '\n')
print('tf \n', tf, '\n')

df = np.count_nonzero(wcX, axis=0) 
print('df: ', df, '\n') 

idf = np.log((1 + N) / (1 + df) )+ 1  if smooth_idf else np.log( N / df )  
print('idf :', idf, '\n')

tfidf = normalize(tf*idf) if norm_idf else tf*idf

df_tfidf = pd.DataFrame(tfidf, columns=wc.get_feature_names())

print(df_tfidf)

4文コーパス: 
 ['This is the first document.', 'This document is the second document.', 'And this is the third one.', 'Is this the first document?'] 

全文書数：  4 

wcX[i, :] 
 [array([0, 1, 1, 1, 0, 0, 1, 0, 1]), array([0, 2, 0, 1, 0, 1, 1, 0, 1]), array([1, 0, 0, 1, 1, 0, 1, 1, 1]), array([0, 1, 1, 1, 0, 0, 1, 0, 1])] 

words_freq 
 [5, 6, 6, 5] 

tf 
 [[0.         0.2        0.2        0.2        0.         0.
  0.2        0.         0.2       ]
 [0.         0.33333333 0.         0.16666667 0.         0.16666667
  0.16666667 0.         0.16666667]
 [0.16666667 0.         0.         0.16666667 0.16666667 0.
  0.16666667 0.16666667 0.16666667]
 [0.         0.2        0.2        0.2        0.         0.
  0.2        0.         0.2       ]] 

df:  [1 3 2 4 1 1 4 1 4] 

idf : [1.91629073 1.22314355 1.51082562 1.         1.91629073 1.91629073
 1.         1.91629073 1.        ] 

        and  document     first        is       one    second       the  \
0  0.000000  0.469791  0.580286  0.384085  0

/usr/local/lib/python3.8/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In [82]:
def getCharacteristicWord():
  result = [[] for i in range(len(df_tfidf["and"]))]

  for column in df_tfidf.columns:
    for i in range(len(df_tfidf[column])):
      if (df_tfidf[column].iloc[i] > 0.45):
        result[i].append(column)

  for i in range(len(result)):
    print("文書",i,result[i])

In [80]:
getCharacteristicWord()

文書 0 ['document', 'first']
文書 1 ['document', 'second']
文書 2 ['and', 'one', 'third']
文書 3 ['document', 'first']


文書0〜3の特徴のある単語を以下に示せ． 𝑡𝑓⋅𝑖𝑑𝑓のしきい値は 𝑡𝑓⋅𝑖𝑑𝑓>0.45 ．　複数ある場合は，カンマ', 'で区切り示してください．

- 文書0: document, first
- 文書1: document, second 
- 文書2: and, one, third
- 文書3: document, first

**問：** 例えば，WikiPediaのスポーツカテゴリの記事中の特徴ある単語を判別する際の指標にTf-Idfを用いる場合，Tf-Idfを算出した後、自動的に特徴のある単語、特徴のない単語を分類するにはどのような方法が考えられるでしょうか。以下の選択肢から記号を選んでください。  (複数選択可)

A: Tf-Idf値を無視して自分の興味のある単語を選ぶ  
B: 閾値を調整し、閾値の範囲で決定する  
C: 結果（単語とTf-Idf値）を目視で調査し，上位何位までを特徴のある単語とする <br>
D: 算出したTf-Idf値を全て採用する

**解答：**
B, C, D

<div style="text-align: right;">
[]点
</div>

---

配点：各20点